Segment v2 use string's real length(#1943)

wangbo opened the pull request:
#1943

wangbo committed:
segment v2 use string's real length(#1943)

imay commented on be/src/olap/field.h:
```diff
@@ -55,9 +55,11 @@ class Field {
 
-    inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
+    virtual inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
+    //todo delete this method when segment v2 release
```
```suggestion
    // TODO(wangbo) delete this method when segment v2 release
```
Should comment why this function stay here to let others know about this.

wangbo replied:
👌

imay commented on be/src/olap/field.h:
```diff
@@ -400,6 +411,16 @@ class VarcharField: public Field {
         return new VarcharField(*this);
     }
+
+    char* allocate_value_from_arena(Arena* arena) const {
```
```suggestion
    char* allocate_value_from_arena(Arena* arena) const override {
```

wangbo replied:
👌

imay commented on be/src/olap/field.h:
```diff
@@ -400,6 +411,16 @@ class VarcharField: public Field {
+        return Field::allocate_string_value_from_arena(arena);
+    }
+
+    void set_to_max(char* ch) {
```
```suggestion
    void set_to_max(char* ch) override {
```

wangbo replied:
👌

imay commented on be/src/olap/field.h:
```diff
@@ -55,9 +55,11 @@ class Field {
-    inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
+    virtual inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
+    //todo delete this method when segment v2 release
+    inline void set_to_max_v1(char* buf) const { return _type_info->set_to_max(buf); }
```
I think there is no need to keep this function, implement it the same with set_to_max is OK. Because the value set is greater than what is set in TypeInfo.

wangbo replied:
implement it the same with set_to_max means:
 when calling varchar field.set_to_max,using field._length to memset slice's data.
 but for segment v1, field._length is regarded as slice.size+sizeof(StringLengthType)
So when calling segment v1 varchar field.set_max,the length of memset is beyond slice's actually size,this may cause subsequently tcmalloc core dumped.see error stack below:
>#0  SLL_Next (t=0xffffffffffffffff) at src/linked_list.h:45
>#1  SLL_TryPop (rv=<synthetic pointer>, list=0x436dce0) at src/linked_list.h:69
>#2  TryPop (rv=<synthetic pointer>, this=0x436dce0) at src/thread_cache.h:220
>#3  Allocate (oom_handler=0x25e5700 <tcmalloc::cpp_throw_oom(unsigned long)>, cl=5, size=64, this=<optimized out>) at src/thread_cache.h:379
>#4  malloc_fast_path<tcmalloc::cpp_throw_oom> (size=64) at src/tcmalloc.cc:1855
>#5  tc_new (size=64) at src/tcmalloc.cc:1976
>#6  0x0000000000c21672 in doris::WrapperField::WrapperField (this=0x4e07d40, rep=<optimized out>, variable_len=<optimized out>, 
>    is_string_type=<optimized out>) at /home/wangbo/incubator-doris/be/src/olap/wrapper_field.cpp:84
>#7  0x0000000000c21b9e in doris::WrapperField::create (column=..., len=<optimized out>, len@entry=0)
>    at /home/wangbo/incubator-doris/be/src/olap/wrapper_field.cpp:51
>#8  0x0000000000cb88e1 in doris::ColumnDataWriter::init (this=0x4bc4780) at /home/wangbo/incubator-doris/be/src/olap/rowset/column_data_writer.cpp:72
>#9  0x0000000000ca273c in doris::AlphaRowsetWriter::_init (this=this@entry=0x4bf83c0)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:275
>#10 0x0000000000ca3694 in doris::AlphaRowsetWriter::init (this=0x4bf83c0, rowset_writer_context=...)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:73
>#11 0x0000000000c91b03 in doris::RowsetFactory::create_rowset_writer (context=..., output=output@entry=0x7fffcdf75318)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/rowset_factory.cpp:50
>#12 0x0000000000bc9b3f in doris::TabletManager::_create_inital_rowset (this=this@entry=0x4d5a180, tablet=..., request=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:1250
>#13 0x0000000000bcc919 in doris::TabletManager::_internal_create_tablet (this=this@entry=0x4d5a180, alter_type=alter_type@entry=doris::SCHEMA_CHANGE, 
>    request=..., is_schema_change_tablet=is_schema_change_tablet@entry=false, ref_tablet=..., data_dirs=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:371
>#14 0x0000000000bcdd72 in doris::TabletManager::create_tablet (this=this@entry=0x4d5a180, request=..., stores=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:318
>#15 0x0000000000b82642 in doris::StorageEngine::create_tablet (this=<optimized out>, request=...)
>    at /home/wangbo/incubator-doris/be/src/olap/storage_engine.cpp:760
>#16 0x0000000000b4fac3 in doris::TestDeleteConditionHandler::SetUp (this=0x4e08460)
>    at /home/wangbo/incubator-doris/be/test/olap/delete_handler_test.cpp:173
>#17 0x0000000002530f37 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#18 0x000000000252ba9c in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#19 0x0000000002510df7 in testing::Test::Run() ()
>#20 0x00000000025116ce in testing::TestInfo::Run() ()
>#21 0x0000000002511d25 in testing::TestCase::Run() ()
>#22 0x0000000002518899 in testing::internal::UnitTestImpl::RunAllTests() ()
>#23 0x0000000002531eb3 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#24 0x000000000252c7a2 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#25 0x0000000002517570 in testing::UnitTest::Run() ()
>#26 0x00000000009fb316 in RUN_ALL_TESTS () at /home/wangbo/incubator-doris/thirdparty/installed/include/gtest/gtest.h:2233
>#27 main (argc=<optimized out>, argv=0x7fffcdf75f88) at /home/wangbo/incubator-doris/be/test/olap/delete_handler_test.cpp:951

it seems that memset memory to 0xff may effect tcmalloc logic

wangbo replied:
when memset a char array to 0xff,if the length of memset is gt than the char array's real length,this may cause tcmalloc core dump,So I think it's better to memset using char array's real length. core dumped as below:

>#0  SLL_Next (t=0x985ca2758b7b4b00) at src/linked_list.h:45
>#1  SLL_TryPop (rv=<synthetic pointer>, list=0x403ece0) at src/linked_list.h:69
>#2  TryPop (rv=<synthetic pointer>, this=0x403ece0) at src/thread_cache.h:220
>#3  Allocate (oom_handler=0x25e8f00 <tcmalloc::cpp_throw_oom(unsigned long)>, cl=5, size=64, this=<optimized out>) at src/thread_cache.h:379
>#4  malloc_fast_path<tcmalloc::cpp_throw_oom> (size=size@entry=64) at src/tcmalloc.cc:1855
>#5  tc_new (size=size@entry=64) at src/tcmalloc.cc:1976
>#6  0x000000000119ba04 in doris::RowBlock::RowBlock (this=0x4894820, schema=<optimized out>)
>    at /home/wangbo/incubator-doris/be/src/olap/row_block.cpp:44
>#7  0x0000000000cbc5ad in doris::ColumnDataWriter::init (this=0x4894780) at /home/wangbo/incubator-doris/be/src/olap/rowset/column_data_writer.cpp:82
>#8  0x0000000000ca5d1c in doris::AlphaRowsetWriter::_init (this=this@entry=0x48c8500)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:275
>#9  0x0000000000ca71b2 in doris::AlphaRowsetWriter::init (this=0x48c8500, rowset_writer_context=...)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/alpha_rowset_writer.cpp:73
>#10 0x0000000000c94df3 in doris::RowsetFactory::create_rowset_writer (context=..., output=output@entry=0x7fff27799208)
>    at /home/wangbo/incubator-doris/be/src/olap/rowset/rowset_factory.cpp:50
>#11 0x0000000000bc711a in doris::TabletManager::_create_inital_rowset (this=this@entry=0x4a2a180, tablet=..., request=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:1250
>#12 0x0000000000bca139 in doris::TabletManager::_internal_create_tablet (this=this@entry=0x4a2a180, alter_type=alter_type@entry=doris::SCHEMA_CHANGE, 
>    request=..., is_schema_change_tablet=is_schema_change_tablet@entry=false, ref_tablet=..., data_dirs=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:371
>#13 0x0000000000bcb592 in doris::TabletManager::create_tablet (this=this@entry=0x4a2a180, request=..., stores=...)
>    at /home/wangbo/incubator-doris/be/src/olap/tablet_manager.cpp:318
>#14 0x0000000000b7fc32 in doris::StorageEngine::create_tablet (this=<optimized out>, request=...)
>    at /home/wangbo/incubator-doris/be/src/olap/storage_engine.cpp:760
>#15 0x0000000000b4f97c in doris::TestDeleteConditionHandler::SetUp (this=0x4af2540)
>    at /home/wangbo/incubator-doris/be/test/olap/delete_handler_test.cpp:173
>#16 0x0000000002534737 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#17 0x000000000252f29c in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
>#18 0x00000000025145f7 in testing::Test::Run() ()
>#19 0x0000000002514ece in testing::TestInfo::Run() ()
>#20 0x0000000002515525 in testing::TestCase::Run() ()
>#21 0x000000000251c099 in testing::internal::UnitTestImpl::RunAllTests() ()
>#22 0x00000000025356b3 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#23 0x000000000252ffa2 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
>#24 0x000000000251ad70 in testing::UnitTest::Run() ()
>#25 0x00000000009fb5f6 in RUN_ALL_TESTS () at /home/wangbo/incubator-doris/thirdparty/installed/include/gtest/gtest.h:2233
>#26 main (argc=<optimized out>, argv=0x7fff27799e78) at /home/wangbo/incubator-doris/be/test/olap/delete_handler_test.cpp:951

imay replied:
why the length of memset is gt than the char array's real length?

wangbo replied:
> why the length of memset is gt than the char array's real length?

For Varchar field,use set_to_max instead of set_to_max_v1 means using Field._length to set slice.size. For segment v1, Field._length=slice.size + sizeof(StringLengthType)(see WrapperField.create), Field._length > slice.size
when calling set_to_max in segment v1 code,Field._length will be used to memset slice.size,so memset's length will gt than slice.size.

imay replied:
Now, when create table column.length is added by sizeof(StringLengthType). I think it's OK to use (column.length - sizeof(StringLengthType)) whether it is v1 or v2. If we do in that way, this will be OK, set_to_max can be unified.

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/segment_writer.cpp:
```diff
@@ -55,17 +55,17 @@ Status SegmentWriter::init(uint32_t write_mbytes_per_sec) {
         }
-        std::unique_ptr<ColumnWriter> writer(new ColumnWriter(opts, type_info, is_nullable, _output_file.get()));
+
+        Field* field = FieldFactory::create(column);
```
It's not a good way to write code like this, it made me feel that the field had been leaked. I think you can use a std::unique_ptr and std::move to finish it.
```suggestion
        std::unique_ptr<Field> field(FieldFactory::create(column));
```

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/segment_writer.cpp:
```diff
@@ -55,17 +55,17 @@ Status SegmentWriter::init(uint32_t write_mbytes_per_sec) {
+        Field* field = FieldFactory::create(column);
+        DCHECK(field != nullptr);
+
+        std::unique_ptr<ColumnWriter> writer(new ColumnWriter(opts, field, is_nullable, _output_file.get()));
```
```suggestion
        std::unique_ptr<ColumnWriter> writer(new ColumnWriter(opts, std::move(field), is_nullable, _output_file.get()));
```

imay commented on be/src/olap/rowset/segment_v2/column_writer.cpp:
```diff
@@ -72,13 +72,13 @@ class NullBitmapBuilder {
 
 ColumnWriter::ColumnWriter(const ColumnWriterOptions& opts,
-                           const TypeInfo* typeinfo,
+                           Field* field,
```
```suggestion
                           std::unique_ptr<Field>&& field,
```

wangbo replied:
👌

wangbo committed:
update commit

wangbo committed:
remove set_to_max_v1

imay commented on be/src/olap/field.h:
```diff
@@ -55,9 +55,9 @@ class Field {
+    virtual inline void set_to_max(char* buf) const { return _type_info->set_to_max(buf); }
     inline void set_to_min(char* buf) const { return _type_info->set_to_min(buf); }
-    inline char* allocate_value_from_arena(Arena* arena) const { return _type_info->allocate_value_from_arena(arena); }
+    virtual inline char* allocate_value_from_arena(Arena* arena) const { return _type_info->allocate_value_from_arena(arena); }
```
I think we can remove allocate_value_from_arena from TypeInfo, instead we can allocate memory here.

wangbo replied:
But Field has no info about sizeof(CppType),only TypeInfo knows it

imay replied:
we can get it through _type_info->size()

wangbo replied:
👌

imay commented on be/src/olap/field.h:
```diff
@@ -389,6 +397,7 @@ class VarcharField: public Field {
         return  _length - OLAP_STRING_MAX_BYTES;
     }
 
+    //todo(wangbo) using _length replace (_length - OLAP_STRING_MAX_BYTES) when segment v2 release
```
I think we can't replace _length, because all _length has been written in tablet metadata. You can comment why we minus OLAP_STRING_MAX_BYTES. Because when v1 is deleted, other people will feel strange why we minus OLAP_STRING_MAX_BYTES.

wangbo replied:
for v2 varchar,If allocate string value'length < field.length,then write a varchar which length is field.length into the allocated value ,this won't cause wrong value?

imay replied:
In your case it would cause problem. But for v2 max length of a varchar field is also field.length - OLAP_STRING_MAX_BYTES. It's impossible to write a longer value to allocated field.

wangbo replied:
👌;allocate_memory is used by max/min value and row_cursor used to write value
I checked latter case, varchar which allocated by Field.allocate_memory is written by direct copy,so the length here doesn't matter

imay commented on be/src/olap/types.h:
```diff
@@ -601,22 +594,11 @@ struct FieldTypeTraits<OLAP_FIELD_TYPE_VARCHAR> : public FieldTypeTraits<OLAP_FI
         slice->size = value_len;
         return OLAP_SUCCESS;
     }
-    static void set_to_max(void* buf) {
```
If you remove this function, it will use CHAR's set_to_max, which memset with slice->size. It is not what we need. I think you should set set_to_max to nullptr to avoid use VARCHAR's set_to_max. What's more I prefer that move CHAR's set_to_max to field too. In CharFiled we can use field's length other than slice->size, which is less error-prone.

wangbo replied:
👌；We can unify char/varchar's set_to_max

wangbo replied:
> set set_to_max to nullptr to avoid use VARCHAR's set_to_max
Here means set a Class's static function to nullptr?

imay replied:
Yes, define a static function pointer who's value is nullptr;

wangbo committed:
1. set type<char/varchar> set_to_max to nullptr
2. change allocate_memory comment
3. remove allocate_value_from_arena from type.h

imay reviewed (approved):
LGTM
